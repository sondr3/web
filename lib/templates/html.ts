/**
 * My handrolled JSX-like solution to templating. Its gloriously awful.
 *
 * @param input - Some HTML
 * @param values - Values that are coerced to strings
 * @returns The converted HTML
 */
export const html = (input: TemplateStringsArray, ...values: unknown[]): string => {
  let output = ""
  input.forEach((string, i) => {
    const value = values[i]

    if (Array.isArray(value)) {
      output += string.toString() + value.join("")
    } else if (typeof value === "string") {
      output += string.toString() + value
    } else if (typeof value === "number") {
      output += string.toString() + value.toString()
    } else {
      output += string
    }
  })

  return output
}
