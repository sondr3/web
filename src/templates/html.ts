/**
 * My handrolled JSX-like solution to templating. Its gloriously awful.
 *
 * @param input - Some HTML
 * @param values - Values that are coerced to strings
 * @returns The converted HTML
 */
export const html = (input: TemplateStringsArray, ...values: readonly unknown[]): string => {
  return input
    .map((string, index) => {
      const value = values[index];

      if (Array.isArray(value)) {
        return string.toString() + value.join("");
      } else if (typeof value === "string") {
        return string.toString() + value;
      } else if (typeof value === "number") {
        return string.toString() + value.toString();
      } else {
        return string;
      }
    })
    .join("");
};
