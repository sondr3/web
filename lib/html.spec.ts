import { html } from "./html";

describe("HTML", () => {
  it("can render som HTML", () => {
    const things = [
      { name: "Test1", bool: false, val: 1 },
      { name: "Test2", bool: true, val: 2 },
      { name: "Test3", bool: true, val: 3 },
      { name: "Test4", bool: true, val: 4 },
    ];

    const document = (body: string, head: string): string =>
      html`<html lang="en">
        ${head}
        <body>
          ${body}
        </body>
      </html>`;

    const header = html`
      <head>
        <title>Hello!</title>
        <meta charset="UTF-8" />
        <meta content="width=device-width, initial-scale=1" name="viewport" />
      </head>
    `;

    const output = document(
      html`<div>
        <p>Hello, world!</p>
        <div>${things.map(({ name, bool, val }) => html`<span>${name} - ${bool.valueOf()}: ${val}</span>`)}</div>
      </div>`,
      header,
    );

    expect(typeof output).toBe("string");
    expect(output).toMatchSnapshot();
  });
});
