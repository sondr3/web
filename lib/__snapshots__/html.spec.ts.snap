// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML can render som HTML 1`] = `
"<html lang=\\"en\\">
        
      <head>
        <title>Hello!</title>
        <meta charset=\\"UTF-8\\" />
        <meta content=\\"width=device-width, initial-scale=1\\" name=\\"viewport\\" />
      </head>
    
        <body>
          <div>
        <p>Hello, world!</p>
        <div><span>Test1 - : 1</span><span>Test2 - : 2</span><span>Test3 - : 3</span><span>Test4 - : 4</span></div>
      </div>
        </body>
      </html>"
`;
